typing/parmatch.ml

```diff
@@ -44,6 +44,132 @@ let omega_list l = List.map (fun _ -> omega) l
 
 let zero = make_pat (Tpat_constant (Const_int 0)) Ctype.none Env.empty
 
+module Simple_head_pat : sig
+  type desc =
+    | Any
+    | Construct of constructor_description
+    | Constant of constant
+    | Tuple of int
+    | Record of label_description list
+    | Variant of { tag: label; has_arg: bool; row: row_desc ref }
+    | Array of int
+    | Lazy
+
+  type t
+
+  val desc : t -> desc
+  val env : t -> Env.t
+  val loc : t -> Location.t
+  val typ : t -> Types.type_expr
+
+  (** [of_pattern p] raises [Invalid_arg _] if [p] is an or-pattern or an
+      exception-pattern. *)
+  val of_pattern : pattern -> t
+
+  (** reconstructs a pattern, putting wildcards as sub-patterns. *)
+  val to_pattern : t -> pattern
+
+  val make_record
+    :  loc:Location.t
+    -> typ:Types.type_expr
+    -> env:Env.t
+    -> label_description list
+    -> t
+
+  val row_of_discr : t -> Types.row_desc
+
+end = struct
+  type desc =
+    | Any
+    | Construct of constructor_description
+    | Constant of constant
+    | Tuple of int
+    | Record of label_description list
+    | Variant of { tag: label; has_arg: bool; row: row_desc ref }
+    | Array of int
+    | Lazy
+
+  type t = {
+    desc: desc;
+    typ : Types.type_expr;
+    loc : Location.t;
+    env : Env.t;
+    attributes : attributes;
+  }
+
+  let desc { desc } = desc
+  let env { env } = env
+  let loc { loc } = loc
+  let typ { typ } = typ
+
+  let row_of_discr d =
+    match Ctype.expand_head d.env d.typ with
+    | {desc = Tvariant row} -> Btype.row_repr row
+    | _ -> assert false
```

Comment:
This function is not very nice. Why don't we just let the user match the `desc`, which should be a `Variant`, and access the `row` value there? (See comment at the use-site below.)

Reply:
There is a two-parts answer to that question:
1. This function already existed: I merely reproduced the behavior of the existing code. So the question should really be directed at the (author of the) existing code. Of course we could take this PR as an opportunity to clean things up, however:
2.  a) I am not sure that the row you will find in the type is the same row that you will find on in `Tpat_variant`. Naively one would hope that they are the same (and there are good reasons to think they are), but I'm just not sure.
  b) In an initial version of the patch, I had made this simplifications, and then either the compiler failed to build itself, or the testsuite was failing (I can't quite remember). It's possible that the simplification was not to blame, that I had made another silly mistake. But the issue just isn't there without the "simplification".

Reply:
Uh, this sounds scary. I understand why you don't want to touch it anymore. Could you maybe try to check once again that using the `Variant`'s row here causes an issue, and then add a comment here to this effect?

Reply:
```
$ make world.opt

...

File ".../utils/profile.ml", line 239, characters 15-424:
239 | ...............(function
240 |         | `Time ->
241 |           make p.duration ~f:time_display
242 |         | `Alloc ->
243 |           make p.allocated_words ~f:memory_word_display
...
246 |         | `Abs_top_heap ->
247 |           make (float_of_int top_heap_words)
248 |            ~f:(memory_word_display ~previous:(float_of_int prev_top_heap_words))
249 |       ).........
Error (warning 8): this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
`AnyOtherTag
```

Clearly, that isn't right.

The patch is here: https://gist.github.com/trefis/d6024cb33c5a84681e449e3741d3b641